client/consensus/epochs/src/lib.rs

```diff
@@ -67,97 +69,209 @@ impl<'a, H, Block> IsDescendentOfBuilder<Block::Hash>
 }
 
 /// Epoch data, distinguish whether it is genesis or not.
+///
+/// Once an epoch is created, it must have a known `start_slot` and `end_slot`, which cannot be
+/// changed. Consensus engine can modify and other data into the epoch, if needed.
 pub trait Epoch {
 	/// Descriptor for the next epoch.
 	type NextEpochDescriptor;
 	/// Type of the slot number.
-	type SlotNumber: Ord;
+	type SlotNumber: Ord + Copy;
 
+	/// The starting slot of the epoch.
+	fn start_slot(&self) -> Self::SlotNumber;
+	/// The end slot of the epoch.
```

Comment:
> /// Produce the "end slot" of the epoch. This is NOT inclusive to the epoch,
> /// i.e. the slots covered by the epoch are `self.start_slot() .. self.end_slot()`. 

Do you think this is not relevant for the generic epoch logic below? The pruning logic relies on this IIRC.

Reply:
This is still relevant to the generic epoch logic, I think. However, I don't believe we have any special logic for genesis `end_slot`. It should be treated the same as any other epochs.

I'll add back the comments.